webpack tries to resolve `new URL('./', import.meta.url)` to a module and fails or succeeds (both are wrong)

<!-- Please don't delete this template because we'll close your issue -->
<!-- Before creating an issue please make sure you are using the latest version of webpack. -->

# Bug report

<!-- Please ask questions on StackOverflow or the webpack Gitter. -->
<!-- https://stackoverflow.com/questions/ask?tags=webpack -->
<!-- https://gitter.im/webpack/webpack -->
<!-- Issues which contain questions or support requests will be closed. -->

**What is the current behavior?**

Webpack's resolution of `new URL(..., import.meta.url)` does something unexpected when the target is a directory, which doesn't correspond to what either the browser or Node will do.  For comparison, try this in $BROWSER_OF_YOUR_CHOICE:
```html
<pre id="output">
</pre>
<script type="module">
  const pre = document.getElementById("output");
  pre.innerHTML += new URL("./", import.meta.url).href;
</script>
```
or in Node, create `index.mjs` and run it:
```js
console.log(new URL("./", import.meta.url).href);
```
In both cases you should see something like:
```
file:///home/dhd/work/webpack-wasm-test/importmeta/
```
Webpack does ... something else.  See below.

**If the current behavior is a bug, please provide the steps to reproduce.**

Given the `index.mjs` file above, run:
```
webpack ./index.mjs
```
It will fail with an error like this:
```
ERROR in ./index.mjs 1:12-42
Module not found: Error: Can't resolve './' in '/home/dhd/work/importmeta'
resolve './' in '/home/dhd/work/importmeta'
  Parsed request is a directory
  No description file found in /home/dhd/work/importmeta or above
  No description file found in /home/dhd/work or above
  as directory
    existing directory /home/dhd/work/importmeta
      No description file found in /home/dhd/work/importmeta or above
      using path: /home/dhd/work/importmeta/index
        No description file found in /home/dhd/work/importmeta or above
        no extension
          /home/dhd/work/importmeta/index doesn't exist
```
(if there is a `package.json` somewhere in a parent directory it will output some other stuff about that, but it will still fail in the same way)

I can make it "work" if I initialize the current directory as a module:
```
npm init
npm install -D webpack-cli webpack
touch index.js
npx webpack ./index.mjs
```
This is *not* what I want, because it will resolve the `new URL` above to the module in the current directory, then bundle whatever it can find in that module.  All I wanted was a URL that pointed to the current directory! (and a Pepsi)

<!-- A great way to do this is to provide your configuration via a GitHub repository -->

See the example here: https://github.com/dhdaines/webpack-wasm-test/tree/main/importmeta

<!-- The most helpful is a minimal reproduction with instructions on how to reproduce -->
<!-- Repositories with too many files or large `webpack.config.js` files are not suitable -->
<!-- Please only add small code snippets directly into this issue -->
<!-- https://gist.github.com is a good place for longer code snippets -->
<!-- If your issue is caused by a plugin or loader, please create an issue on the loader/plugin repository instead -->

**What is the expected behavior?**

Webpack should just pass `new URL('./', import.meta.url)` through unchanged.  It should not try to resolve a directory as a module. 
 Ideally it should also pass through unchanged any other instance of `new URL(..., import.meta.url)` where the target can't be resolved as an asset, since the user might create them after the fact.

<!-- "It should work" is not a helpful explanation -->
<!-- Explain exactly how it should behave -->

**Other relevant information:**
webpack version:  ^5.76.3 => 5.76.3
Node.js version: 14.18.2
Operating System: Linux 5.19 Ubuntu 22.04.2 LTS 22.04.2 LTS (Jammy Jellyfish)

--- a/.changeset/url-import-meta-directory.md
+++ b/.changeset/url-import-meta-directory.md
@@ -1,0 +1,5 @@
+---
+"webpack": patch
+---
+
+Pass through `new URL(...)` directory references instead of resolving them.

--- a/lib/url/URLParserPlugin.js
+++ b/lib/url/URLParserPlugin.js
@@ -56,6 +56,32 @@ const isMetaUrl = (parser, arg) => {
 	return true;
 };
 
+/**
+ * Whether the request points to a directory and thus can't be an asset.
+ * @param {string} request request
+ * @returns {boolean} true when the request is a directory reference
+ */
+const isDirectoryRequest = (request) =>
+	request === "." || request === ".." || request.endsWith("/");
+
+/**
+ * Replaces `import.meta.url` with the runtime base URI, leaving the
+ * `new URL(...)` construction untouched so it's evaluated at runtime.
+ * @param {JavascriptParser} parser parser
+ * @param {NewExpressionNode} expr expression
+ * @returns {void}
+ */
+const keepNewURL = (parser, expr) => {
+	const [, arg2] = expr.arguments;
+	const dep = new ConstDependency(
+		RuntimeGlobals.baseURI,
+		/** @type {Range} */ (arg2.range),
+		[RuntimeGlobals.baseURI]
+	);
+	dep.loc = /** @type {DependencyLocation} */ (expr.loc);
+	parser.state.module.addPresentationalDependency(dep);
+};
+
 /** @type {WeakMap<NewExpressionNode, BasicEvaluatedExpression | undefined>} */
 const getEvaluatedExprCache = new WeakMap();
 
@@ -157,13 +183,7 @@ class URLParserPlugin {
 						return;
 					}
 
-					const dep = new ConstDependency(
-						RuntimeGlobals.baseURI,
-						/** @type {Range} */ (arg2.range),
-						[RuntimeGlobals.baseURI]
-					);
-					dep.loc = /** @type {DependencyLocation} */ (expr.loc);
-					parser.state.module.addPresentationalDependency(dep);
+					keepNewURL(parser, expr);
 
 					return true;
 				}
@@ -177,6 +197,12 @@ class URLParserPlugin {
 
 			// static URL
 			if ((request = evaluatedExpr.asString())) {
+				// A directory is not an asset; keep `new URL(...)` as-is, matching Node/browsers.
+				if (isDirectoryRequest(request)) {
+					keepNewURL(parser, expr);
+					return true;
+				}
+
 				const [arg1, arg2] = expr.arguments;
 				const dep = new URLDependency(
 					request,

--- a/test/configCases/url/import-meta-url-directory/index.js
+++ b/test/configCases/url/import-meta-url-directory/index.js
@@ -1,0 +1,13 @@
+it("should not resolve a directory reference as a module and keep `new URL(...)` as-is", () => {
+	const dotSlash = new URL("./", import.meta.url);
+	expect(dotSlash.href.endsWith("/")).toBe(true);
+
+	const dot = new URL(".", import.meta.url);
+	expect(dot.href.endsWith("/")).toBe(true);
+
+	const parent = new URL("../", import.meta.url);
+	expect(parent.href.endsWith("/")).toBe(true);
+
+	const subDir = new URL("./folder/", import.meta.url);
+	expect(subDir.href.endsWith("/folder/")).toBe(true);
+});

--- a/test/configCases/url/import-meta-url-directory/webpack.config.js
+++ b/test/configCases/url/import-meta-url-directory/webpack.config.js
@@ -1,0 +1,6 @@
+"use strict";
+
+/** @type {import("../../../../").Configuration} */
+module.exports = {
+	target: "web"
+};

